File: inventory.py
import pygame
from typing import Dict, List, Optional, Type

class Inventory:
    """Manages the player's collected items"""
    
    def __init__(self, max_slots=24):
        self.max_slots = max_slots
        self.items = []  # List of items
        self.item_counts = {}  # Dictionary to track quantities of stackable items
    
    def add_item(self, item) -> bool:
        """
        Add an item to the inventory
        
        Args:
            item: The item object to add
            
        Returns:
            bool: True if successfully added, False if inventory is full
        """
        # Check if inventory is full
        if len(self.items) >= self.max_slots and item.name not in self.item_counts:
            print(f"Inventory full, cannot add {item.name}")
            return False
        
        # Check if the item is stackable and we already have one
        if hasattr(item, 'stackable') and item.stackable and item.name in self.item_counts:
            # Increase count for existing item
            self.item_counts[item.name] += 1
            print(f"Added {item.name} to stack (now have {self.item_counts[item.name]})")
            return True
        
        # Add as new item
        self.items.append(item)
        # Initialize count for stackable items
        if hasattr(item, 'stackable') and item.stackable:
            self.item_counts[item.name] = 1
        
        print(f"Added {item.name} to inventory (slot {len(self.items)-1})")
        return True
    
    def remove_item(self, item_index) -> Optional[object]:
        """
        Remove an item from inventory by index
        
        Args:
            item_index: Index of item to remove
            
        Returns:
            The removed item or None if index is invalid
        """
        if 0 <= item_index < len(self.items):
            item = self.items[item_index]
            
            # Check if it's a stacked item
            if hasattr(item, 'stackable') and item.stackable and item.name in self.item_counts:
                if self.item_counts[item.name] > 1:
                    # Reduce stack count
                    self.item_counts[item.name] -= 1
                    print(f"Removed one {item.name} from stack (remaining: {self.item_counts[item.name]})")
                    return item
                else:
                    # Remove last item from stack
                    del self.item_counts[item.name]
            
            # Remove item from inventory
            self.items.pop(item_index)
            print(f"Removed {item.name} from inventory")
            return item
        
        return None
    
    def get_item_by_name(self, item_name) -> Optional[object]:
        """
        Find the first item with the given name
        
        Args:
            item_name: Name of the item to find
            
        Returns:
            The found item or None if not found
        """
        for item in self.items:
            if item.name == item_name:
                return item
        return None
    
    def get_item_count(self, item_name) -> int:
        """
        Get the count of a specific item
        
        Args:
            item_name: Name of the item to count
            
        Returns:
            int: The count of the item (0 if none found)
        """
        return self.item_counts.get(item_name, 0)
    
    def get_all_items(self) -> List:
        """
        Get all items in the inventory
        
        Returns:
            List of all items
        """
        return self.items.copy()
    
    def get_items_and_counts(self) -> List[Dict]:
        """
        Get all items with their counts
        
        Returns:
            List of dictionaries with item and count information
        """
        result = []
        for item in self.items:
            count = 1
            if hasattr(item, 'stackable') and item.stackable and item.name in self.item_counts:
                count = self.item_counts[item.name]
            
            result.append({
                'item': item,
                'count': count
            })
        
        return result
    
    def clear(self):
        """Clear the entire inventory"""
        self.items = []
        self.item_counts = {}
        print("Inventory cleared")
    
    def is_full(self) -> bool:
        """Check if inventory is full"""
        return len(self.items) >= self.max_slots
    
    def get_empty_slots(self) -> int:
        """Get number of empty slots"""
        return max(0, self.max_slots - len(self.items))

File: src/entities/inventory.js
/**
 * Inventory - Manages the player's collected items
 * Mirrors Python inventory.py
 */
class Inventory {
    constructor(maxSlots = 24) {
        this.maxSlots = maxSlots;
        this.items = [];  // List of items
        this.itemCounts = {};  // Dictionary to track quantities of stackable items
    }

    /**
     * Add an item to the inventory
     * @param {Object} item - The item object to add
     * @returns {boolean} True if successfully added, False if inventory is full
     */
    addItem(item) {
        // Check if inventory is full
        if (this.items.length >= this.maxSlots && !this.itemCounts[item.name]) {
            console.log(`Inventory full, cannot add ${item.name}`);
            return false;
        }

        // Check if the item is stackable and we already have one
        if (item.stackable && this.itemCounts[item.name]) {
            // Increase count for existing item
            this.itemCounts[item.name]++;
            console.log(`Added ${item.name} to stack (now have ${this.itemCounts[item.name]})`);
            return true;
        }

        // Add as new item
        this.items.push(item);
        // Initialize count for stackable items
        if (item.stackable) {
            this.itemCounts[item.name] = 1;
        }

        console.log(`Added ${item.name} to inventory (slot ${this.items.length - 1})`);
        return true;
    }

    /**
     * Remove an item from inventory by index
     * @param {number} itemIndex - Index of item to remove
     * @returns {Object|null} The removed item or null if index is invalid
     */
    removeItem(itemIndex) {
        if (itemIndex >= 0 && itemIndex < this.items.length) {
            const item = this.items[itemIndex];

            // Check if it's a stacked item
            if (item.stackable && this.itemCounts[item.name]) {
                if (this.itemCounts[item.name] > 1) {
                    // Reduce stack count
                    this.itemCounts[item.name]--;
                    console.log(`Removed one ${item.name} from stack (remaining: ${this.itemCounts[item.name]})`);
                    return item;
                } else {
                    // Remove last item from stack
                    delete this.itemCounts[item.name];
                }
            }

            // Remove item from inventory
            this.items.splice(itemIndex, 1);
            console.log(`Removed ${item.name} from inventory`);
            return item;
        }

        return null;
    }

    /**
     * Find the first item with the given name
     * @param {string} itemName - Name of the item to find
     * @returns {Object|null} The found item or null if not found
     */
    getItemByName(itemName) {
        for (const item of this.items) {
            if (item.name === itemName) {
                return item;
            }
        }
        return null;
    }

    /**
     * Get the count of a specific item
     * @param {string} itemName - Name of the item to count
     * @returns {number} The count of the item (0 if none found)
     */
    getItemCount(itemName) {
        return this.itemCounts[itemName] || 0;
    }

    /**
     * Get all items in the inventory
     * @returns {Array} List of all items
     */
    getAllItems() {
        return [...this.items];
    }

    /**
     * Get all items with their counts
     * @returns {Array} List of objects with item and count information
     */
    getItemsAndCounts() {
        const result = [];
        for (const item of this.items) {
            let count = 1;
            if (item.stackable && this.itemCounts[item.name]) {
                count = this.itemCounts[item.name];
            }

            result.push({
                item: item,
                count: count
            });
        }
        return result;
    }

    /**
     * Clear the entire inventory
     */
    clear() {
        this.items = [];
        this.itemCounts = {};
        console.log('Inventory cleared');
    }

    /**
     * Check if inventory is full
     * @returns {boolean}
     */
    isFull() {
        return this.items.length >= this.maxSlots;
    }

    /**
     * Get number of empty slots
     * @returns {number}
     */
    getEmptySlots() {
        return Math.max(0, this.maxSlots - this.items.length);
    }
}

// Export
window.Inventory = Inventory;
